src/stream/double_ended/mod.rs

```diff
@@ -0,0 +1,270 @@
+
+            fn increment(
+                s: impl DoubleEndedStream<Item = i32> + Unpin,
+            ) -> impl DoubleEndedStream<Item = i32> + Unpin {
+                struct Increment<S>(S);
+
+                impl<S: DoubleEndedStream<Item = i32> + Unpin> Stream for Increment<S> {
+                    type Item = S::Item;
+
+                    fn poll_next_back(
+                        mut self: Pin<&mut Self>,
+                        cx: &mut Context<'_>,
+                    ) -> Poll<Option<Self::Item>> {
+                        match Pin::new(&mut self.0).poll_next_back(cx) {
+                            Poll::Pending => Poll::Pending,
+                            Poll::Ready(None) => Poll::Ready(None),
+                            Poll::Ready(Some(item)) => Poll::Ready(Some(item + 1)),
+                        }
+                    }
+                }
+
+                Increment(s)
+            }
+
+            let mut s = increment(stream::once(7)); // will need to implement DoubleEndedStream
+
+            assert_eq!(s.next_back().await, Some(8));
+            assert_eq!(s.next_back().await, None);
+            #
+            # }) }
+            ```
+        "#]
+        fn poll_next_back(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>>;
+    }
+
+    #[doc = r#"
+        Extension methods for [`DoubleEndedStreamExt`].
+
+        [`Stream`]: ../stream/trait.Stream.html
+    "#]
+    pub trait DoubleEndedStreamExt: crate::stream::DoubleEndedStream {
+        #[doc = r#"
+            Advances the stream and returns the next value.
+
+            Returns [`None`] when iteration is finished. Individual stream implementations may
+            choose to resume iteration, and so calling `next()` again may or may not eventually
+            start returning more values.
+
+            [`None`]: https://doc.rust-lang.org/std/option/enum.Option.html#variant.None
+
+            # Examples
+
+            ```
+            # fn main() { async_std::task::block_on(async {
+            #
+            use async_std::stream::Sample;
+            use async_std::stream::double_ended::DoubleEndedStreamExt;
+
+            let mut s = Sample::from(vec![7u8]);
+
+            assert_eq!(s.next_back().await, Some(7));
+            assert_eq!(s.next_back().await, None);
+            #
+            # }) }
+            ```
+        "#]
+        fn next_back(&mut self) -> impl Future<Output = Option<Self::Item>> + '_ [NextBackFuture<'_, Self>]
+            where
+                Self: Unpin,
+        {
+            NextBackFuture { stream: self }
+        }
+
+        #[doc = r#"
+            Returns the nth element from the back of the stream.
+
+            # Examples
+
+            Basic usage:
+
+            ```
+            # fn main() { async_std::task::block_on(async {
+            #
+            use async_std::stream::Sample;
+            use async_std::stream::double_ended::DoubleEndedStreamExt;
```

Comment:
whether should we re-export it to `async_std::stream::DoubleEndedStreamExt` rather than `async_std::stream::double_ended::DoubleEndedStreamExt`?

Reply:
I still need to figure this out, but its good input :)